pkg/kubecost/cloudcost.go

```diff
@@ -0,0 +1,547 @@
+		CloudCostSets: cloudCostItemSets,
+	}, nil
+}
+
+func (ccsr *CloudCostSetRange) Clone() *CloudCostSetRange {
+	ccsSlice := make([]*CloudCostSet, len(ccsr.CloudCostSets))
+	for i, ccs := range ccsr.CloudCostSets {
+		ccsSlice[i] = ccs.Clone()
+	}
+	return &CloudCostSetRange{
+		Window:        ccsr.Window.Clone(),
+		CloudCostSets: ccsSlice,
+	}
+}
+
+func (ccsr *CloudCostSetRange) IsEmpty() bool {
+	for _, ccas := range ccsr.CloudCostSets {
+		if !ccas.IsEmpty() {
+			return false
+		}
+	}
+	return true
+}
+
+// Accumulate sums each CloudCostSet in the given range, returning a single cumulative
+// CloudCostSet for the entire range.
+func (ccsr *CloudCostSetRange) Accumulate() (*CloudCostSet, error) {
+	var cloudCostSet *CloudCostSet
+	var err error
+
+	for _, ccs := range ccsr.CloudCostSets {
+		if cloudCostSet == nil {
+			cloudCostSet = ccs.cloneSet()
+			continue
+		}
+		err = cloudCostSet.accumulateInto(ccs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cloudCostSet, nil
+}
+
+// LoadCloudCost loads CloudCosts into existing CloudCostSets of the CloudCostSetRange.
+// This function service to aggregate and distribute costs over predefined windows
+// are accumulated here so that the resulting CloudCost with the 1d window has the correct price for the entire day.
+// If all or a portion of the window of the CloudCost is outside of the windows of the existing CloudCostSets,
+// that portion of the CloudCost's cost will not be inserted
+func (ccsr *CloudCostSetRange) LoadCloudCost(cloudCost *CloudCost) {
+	window := cloudCost.Window
+	if window.IsOpen() {
+		log.Errorf("CloudCostSetRange: LoadCloudCost: invalid window %s", window.String())
+		return
+	}
+
+	totalPct := 0.0
+
+	// Distribute cost of the current item across one or more CloudCosts in
+	// across each relevant CloudCostSet. Stop when the end of the current
+	// block reaches the item's end time or the end of the range.
+	for _, ccs := range ccsr.CloudCostSets {
+		setWindow := ccs.Window
+
+		// get percent of item window contained in set window
+		pct := setWindow.GetPercentInWindow(window)
+		if pct == 0 {
+			continue
+		}
+
+		cc := cloudCost
+		// If the current set Window only contains a portion of the CloudCost Window, insert costs relative to that portion
+		if pct < 1.0 {
+			cc = &CloudCost{
+				Properties:       cloudCost.Properties,
+				Window:           window.Contract(setWindow),
+				ListCost:         cloudCost.ListCost.percent(pct),
+				NetCost:          cloudCost.NetCost.percent(pct),
+				AmortizedNetCost: cloudCost.AmortizedNetCost.percent(pct),
+				InvoicedCost:     cloudCost.InvoicedCost.percent(pct),
+			}
+		}
+
+		err := ccs.Insert(cc)
+		if err != nil {
+			log.Errorf("CloudCostSetRange: LoadCloudCost: failed to load CloudCost with window %s: %s", setWindow.String(), err.Error())
+		}
+
+		// If all cost has been inserted then finish
+		totalPct += pct
+		if totalPct >= 1.0 {
+			return
+		}
+	}
+}
```

Comment:
This is an early return condition.
Say we have a range with 3 days but the asset is all on the first day. totalPct will be 1.0 but it will continue to check the window for the days later in the range.

Reply:
I updated the comment, so that is more clear.